File: src/CSharp/DynamicBuffers.CSharp/TypedMessage.cs
using System;
using System.Collections.Generic;
using System.Dynamic;
using System.Linq;
using System.Text;
using Google.Protobuf;
using Google.Protobuf.Reflection;
using Google.Protobuf.WellKnownTypes;

namespace DynamicBuffers;


public partial class TypedMessage
{
    private sealed class DescriptorComparer : IEqualityComparer<MessageDescriptor>
    {
        public static DescriptorComparer Instance { get; } = new ();

        public bool Equals(MessageDescriptor x, MessageDescriptor y)
            => x?.FullName == y?.FullName;

        public int GetHashCode(MessageDescriptor obj)
            => obj?.FullName.GetHashCode() ?? 0;
    }


    /// <summary>
    /// Creates a new instance of <see cref="TypedMessage"/> with specified content.
    /// </summary>
    /// <param name="content"></param>
    /// <param name="registry"></param>
    /// <returns></returns>
    /// <exception cref="NotSupportedException"></exception>
    public static TypedMessage Create(IMessage content, TypeRegistry? registry = null)
    {
        static void visitDefinition(HashSet<MessageDescriptor> metadata, MessageDescriptor descriptor)
        {
            if (!metadata.Add(descriptor))
            {
                return;
            }
            foreach (var field in descriptor.Fields.InFieldNumberOrder())
            {
                if (field.FieldType == FieldType.Message)
                {
                    visitDefinition(metadata, field.MessageType);
                }
            }
        }

        static void visitInstance(HashSet<MessageDescriptor> metadata, TypeRegistry registry, IMessage message)
        {
            if (message is null)
            {
                return;
            }
            if (message is Any any)
            {
                metadata.Add(Any.Descriptor);
                visitInstance(metadata, registry, any.Unpack(registry));
                return;
            }

            visitDefinition(metadata, message.Descriptor);
            foreach (var field in message.Descriptor.Fields.InFieldNumberOrder())
            {
                if (field.IsMap)
                {
                    throw new NotSupportedException("map field is not supported.");
                }
                switch (field.FieldType)
                {
                    case FieldType.Message:
                    case FieldType.Group:
                        if(field.IsRepeated)
                        {
                            var children = (IReadOnlyList<IMessage>)field.Accessor.GetValue(message);
                            foreach (var child in children)
                            {
                                visitInstance(metadata, registry, child);
                            }
                        }
                        else
                        {
                            var child = (IMessage)field.Accessor.GetValue(message);
                            visitInstance(metadata, registry, child);
                        }
                        continue;
                    default:
                        continue;
                }
            }
        }

        var metadata = new HashSet<MessageDescriptor>(DescriptorComparer.Instance);
        visitInstance(metadata, registry ?? TypeRegistry.Empty, content);
        var retval = new TypedMessage();
        retval.Metadata.AddRange(metadata.Select(descr => descr.ToModifiedProto()));
        retval.Content = Any.Pack(content);
        return retval;
    }


    /// <summary>
    /// Gets dynamic access surface for this <see cref="TypedMessage"/> instance.
    /// </summary>
    /// <returns></returns>
    /// <exception cref="NotImplementedException"></exception>
    public dynamic AsDynamic()
    {
        throw new NotImplementedException();
    }
}

File: src/Python/dynamicbuffers/typed_message.py
#!/usr/bin/env python3
"""
Extensions for TypedMessage
"""
from typing import Iterable, Mapping

import google.protobuf.message as _pbm
import google.protobuf.descriptor as _pbd
import google.protobuf.any_pb2 as _pb_any
from ._helpers import to_modified_proto
from .proto.TypedMessage_pb2 import TypedMessage as _TypedMessage  # type: ignore # noqa


_FieldNumberMap = Mapping[int, _pbd.FieldDescriptor]


def _unpack(
        any_: _pb_any.Any,
        type_registry: frozenset[type]) -> _pbm.Message | None:
    for t in type_registry:
        msg = t()
        if any_.Is(msg.DESCRIPTOR):
            any_.Unpack(msg)
            return msg
    return None


def _TypedMessage__create(
        content: _pbm.Message,
        type_registry: frozenset[type] | None = None) -> _TypedMessage:
    def visit_definition(
            metadata: set[_pbd.Descriptor],
            descriptor: _pbd.Descriptor) -> None:
        if descriptor in metadata:
            return
        metadata.add(descriptor)
        fields: _FieldNumberMap = descriptor.fields_by_number
        for field_id in fields:
            field = fields[field_id]
            if field.type == _pbd.FieldDescriptor.TYPE_MESSAGE:
                visit_definition(metadata, field.message_type)

    def visit_instance(
            metadata: set[_pbd.Descriptor],
            type_registry: frozenset[type],
            message: _pbm.Message | None) -> None:
        if message is None:
            return
        if isinstance(message, _pb_any.Any):
            metadata.add(_pb_any.Any.DESCRIPTOR)
            any_content = _unpack(message, type_registry)
            visit_instance(metadata, type_registry, any_content)
            return
        visit_definition(metadata, message.DESCRIPTOR)
        fields: _FieldNumberMap = \
            message.DESCRIPTOR.fields_by_number
        for field_id in fields:
            field = fields[field_id]
            if (field.label == _pbd.FieldDescriptor.LABEL_REPEATED and
                    field.message_type and
                    field.message_type.GetOptions().map_entry):
                raise TypeError('map field is not supported.')
            if field.type == _pbd.FieldDescriptor.TYPE_MESSAGE:
                if field.label == _pbd.FieldDescriptor.LABEL_REPEATED:
                    children: Iterable[_pbm.Message] = \
                        getattr(message, field.name)
                    for child in children:
                        visit_instance(metadata, type_registry, child)
                else:
                    child = getattr(message, field.name)
                    visit_instance(metadata, type_registry, child)
            else:
                continue

    metadata: set[_pbd.Descriptor] = set()
    if type_registry is None:
        type_registry = frozenset()
    visit_instance(metadata, type_registry, content)
    retval = _TypedMessage()
    for descr in metadata:
        retval.metadata.append(to_modified_proto(descr))
    retval.content.Pack(content)
    return retval


_TypedMessage.create = _TypedMessage__create
